octopusdeploy_framework/resource_team_mappers.go

```diff
@@ -0,0 +1,265 @@
+				"display_name":        types.StringValue(group.DisplayName),
+				"display_id_and_name": types.BoolValue(group.DisplayIDAndName),
+			},
+		)
+		groupList[i] = objectValue
+	}
+
+	objectType := types.ObjectType{
+		AttrTypes: map[string]attr.Type{
+			"id":                  types.StringType,
+			"display_name":        types.StringType,
+			"display_id_and_name": types.BoolType,
+		},
+	}
+	return types.ListValueMust(objectType, groupList)
+}
+
+func mapUserRoleSetStateToResource(ctx context.Context, team *teams.Team, userRoles types.Set) []*userroles.ScopedUserRole {
+	if userRoles.IsNull() || userRoles.IsUnknown() {
+		return nil
+	}
+
+	userRoleSetElements := userRoles.Elements()
+	scopedUserRoles := make([]*userroles.ScopedUserRole, len(userRoleSetElements))
+
+	for i, userRoleSetElement := range userRoleSetElements {
+		userRoleSetElementAttributes := userRoleSetElement.(types.Object).Attributes()
+
+		userRoleId := userRoleSetElementAttributes["user_role_id"].(types.String).ValueString()
+		spaceId := userRoleSetElementAttributes["space_id"].(types.String).ValueString()
+
+		scopedUserRole := userroles.NewScopedUserRole(userRoleId)
+		scopedUserRole.TeamID = team.ID
+		scopedUserRole.SpaceID = spaceId
+
+		if id, ok := userRoleSetElementAttributes["id"]; ok && !id.IsNull() && !id.IsUnknown() {
+			scopedUserRole.ID = id.(types.String).ValueString()
+		}
+
+		if envIds, ok := userRoleSetElementAttributes["environment_ids"]; ok && !envIds.IsNull() && !envIds.IsUnknown() {
+			scopedUserRole.EnvironmentIDs = expandStringSet(envIds.(types.Set))
+		}
+
+		if projGroupIds, ok := userRoleSetElementAttributes["project_group_ids"]; ok && !projGroupIds.IsNull() && !projGroupIds.IsUnknown() {
+			scopedUserRole.ProjectGroupIDs = expandStringSet(projGroupIds.(types.Set))
+		}
+
+		if projIds, ok := userRoleSetElementAttributes["project_ids"]; ok && !projIds.IsNull() && !projIds.IsUnknown() {
+			scopedUserRole.ProjectIDs = expandStringSet(projIds.(types.Set))
+		}
+
+		if tenantIds, ok := userRoleSetElementAttributes["tenant_ids"]; ok && !tenantIds.IsNull() && !tenantIds.IsUnknown() {
+			scopedUserRole.TenantIDs = expandStringSet(tenantIds.(types.Set))
+		}
+
+		scopedUserRoles[i] = scopedUserRole
+	}
+
+	return scopedUserRoles
+}
+
+func mapUserRoleSetResourceToState(ctx context.Context, userRoles []*userroles.ScopedUserRole, model types.Set) types.Set {
+	if len(userRoles) == 0 {
+		objectType := types.ObjectType{
+			AttrTypes: map[string]attr.Type{
+				"id":                types.StringType,
+				"user_role_id":      types.StringType,
+				"space_id":          types.StringType,
+				"team_id":           types.StringType,
+				"environment_ids":   types.SetType{ElemType: types.StringType},
+				"project_group_ids": types.SetType{ElemType: types.StringType},
+				"project_ids":       types.SetType{ElemType: types.StringType},
+				"tenant_ids":        types.SetType{ElemType: types.StringType},
+			},
```

Comment:
Can this be centralised somewhere please, right now it's duplicated three times in this function